### PR TITLE
Use bootstrap-toc for sidebars

### LINKED ISSUE #817: Improve TOC for long articles
Useful for long vignettes. The sidebar ToC gets messy for long vignettes.

selection_017

Can be triggered by adding the following block in `content-article.html`:

```html
$if(toc)$
<div id="toc">
<h3>Table of Contents</h3>
$toc$
</div>
$endif$
```

However, I would prefer to have a separate flag separating the ToC inline and in the navbar (in case someone just wants to have the navbar one).

Currently (I guess) its triggered by the `toc` `rmarkdown` flag when the header looks like this:

````r
---
title: "Resampling"
pkgdown:
  as_is: true
output: 
  rmarkdown::html_document:
    toc: true
---
````

I don't know if its possible to add an argument to `build_site()` which then triggers a custom html variable in `content-article.html`.

### LINKED ISSUE #1077: Multi-level table of contents
It seems to be a good idea to visually display the document's nested structure created with headers of different order (h1, h2, etc.). Maybe [this comment](https://github.com/r-lib/pkgdown/issues/817#issuecomment-438051320) was about this feature, but I decided to create a separate issue.
Current pkgdown version can include headers of different depth, but they all "blend" together, which makes it harder to visually see an article's structure.

I recently discovered a way to deal with this situation:

- Post-process created html pages with ToC by adding "toc-lev{x}" class to ToC elements, where "{x}" represents the level of entry (1 is the highest; second is nested into first; etc.). Solution is a little bit "hacky", which uses the way ToC is currently created for articles with nested sections:

<details><summary>Code for modification of one html file based on its path</summary>

```r
add_toc_levels <- function(path) {
  page <- xml2::read_html(path)

  toc_items <- page %>%
    # Find table of contents by id
    xml2::xml_find_first(xpath = "//*[@id='tocnav']") %>%
    # Find all <li> (list elements) with "linkable" descendants which are
    # assumed to represent TOC entries
    xml2::xml_find_all(xpath = "descendant::li[descendant::a[@href]]")

  # Compute levels of entries by counting number of ancestors with class "nav-stacked":
  # they represent headers, into which current entry is nested
  toc_level <- vapply(toc_items, function(item) {
    xml2::xml_find_all(
      item, xpath = "ancestor::*[contains(@class, 'nav-stacked')]"
    ) %>%
      length()
  }, numeric(1))

  toc_class <- paste0("toc-lev", toc_level)

  # Add appropriate classes
  lapply(seq_along(toc_items), function(i) {
    xml2::xml_set_attr(toc_items[i], attr = "class", value = toc_class[i])
  })

  xml2::write_html(page, path)
}
```
</details>

- Add CSS rules for those classes to visually separate different levels. For example, `padding-left: 20px;` for "toc-lev2", `padding-left: 40 px;` for "toc-lev3", etc. This will create "tabbed" structure.

If you think this solution can live in pkgdown, let me know and I'll gladly make a PR.

### LINKED ISSUE #1254: TOC highlighting broke
TOC sections are no longer highlighted during scroll on https://pkgdown.r-lib.org/dev/articles/pkgdown.html

### PR DESCRIPTION
**Pull request by jayhesselberth**
Closes #1077. Closes #1254. Closes #817.

### CODE REVIEW

**Comment by jayhesselberth** [comment]
Preview at https://amazing-mirzakhani-5ff563.netlify.com/index.html

TOCs for reference index/pages and articles are much improved.

The major change is on the news page, where I had to move the CRAN release under the h1 element, otherwise the date ends up in the sidebar and looks cluttered.

File: inst/templates/head.html [hunk 1]
```diff
@@ -23,6 +23,8 @@
 {{/yaml}}
 
 <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>
+<link rel="stylesheet" href="https://cdn.rawgit.com/afeld/bootstrap-toc/v0.4.1/dist/bootstrap-toc.min.css">
+<script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v0.4.1/dist/bootstrap-toc.min.js"></script>
```
Comment (Bisaloo) [inline inst/templates/head.html]: Would it be possible to add a subresource integrity hash or to locally host these files? I don't feel super comfortable running scripts for an unknown github account, who could change their content to something malicious anytime.

File: R/build-news.R [hunk 2]
```diff
@@ -249,13 +249,13 @@ tweak_news_heading <- function(x, versions, timeline) {
   date <- timeline$date[match(versions, timeline$version)]
   date_str <- ifelse(is.na(date), "Unreleased", as.character(date))
 
-  date_nodes <- paste(" <small>", date_str, "</small>", collapse = "") %>%
+  date_nodes <- paste(" <h6 data-toc-skip> CRAN release: ", date_str, "</h6>", collapse = "") %>%
```
Comment (hadley) [inline R/build-news.R]: I'd prefer to keep the existing style — we can use `<h2 data-toc-text="Short text">Longer text</h2>` to customise what appears in the TOC

File: inst/assets/pkgdown.css [hunk 3]
```diff
@@ -168,6 +172,29 @@ a.anchor {
   margin-bottom: 0.5em;
 }
 
+/* bootstrap-toc tweaks */
+nav[data-toggle="toc"] .nav > li > a {
+  font-size: 17px;
```
Comment (hadley) [inline inst/assets/pkgdown.css]: This feels too big to me. Can we make it 1.1em? And not bold?

The left padding also feels a bit too big to me — I think the headings should be left aligned with contents. And where does the purple colour come from?

Comment (jayhesselberth) [reply inst/assets/pkgdown.css]: Site with new tweaks at: https://boring-kare-88d4d5.netlify.com/

Comment (batpigandme) [reply inst/assets/pkgdown.css]: Perhaps a happy medium with the adjustments below?
```css
font-weight: 500;
font-size: 1.4rem;
```
I think a _little_ more pop would help.

Comment (jayhesselberth) [reply inst/assets/pkgdown.css]: With more pop: https://competent-feynman-455f34.netlify.com/news/index.html

Comment (batpigandme) [reply inst/assets/pkgdown.css]: I like it! I just spent _way_ too long trying to figure out how to select the submenu items, and couldn't get it quite right (I am not a CSS pro by any measure), but I think those could be a scoch larger, if someone else knows how to scope it right (if not, nbd).

**Comment by jayhesselberth** [comment]
Final? updates: https://modest-ptolemy-687e8c.netlify.com/

**Comment by jayhesselberth** [comment]
@batpigandme do you have a sense for how these changes would affect tidytemplate?

**Comment by hadley** [comment]
I think the subheadings (e.g. "new features") are now a bit small. 

For the main headings, I definitely prefer `weight: 400` and `font-size: 1.5em`, but don't feel obliged to change it if you prefer the current styling.

**Comment by jayhesselberth** [comment]
One more set of tweaks to increase sizes a bit: https://admiring-brahmagupta-25d9d9.netlify.com/

**Pull request merged**